cirq/circuits/circuit.py

```diff
@@ -1258,6 +1258,109 @@ def _json_dict_(self):
     def _from_json_dict_(cls, moments, device, **kwargs):
         return cls(moments, strategy=InsertStrategy.EARLIEST, device=device)
 
+    def tetris_concat(
+        *circuits: 'cirq.AbstractCircuit', stop_at_first_alignment: bool = False
+    ) -> 'cirq.Circuit':
+        """Concatenates circuits while overlapping them if possible.
+
+        Starts with the first circuit (index 0), then iterates over the other
+        circuits while folding them in. To fold two circuits together, they
+        are placed one after the other and then moved inward until any of:
+
+            a) Just before their operations would collide.
+            b) stop_at_first_alignment==True and their ends (or starts) align
+                for the first time.
+            c) Their ends (or starts) align for the last time.
+
+        Beware that this method is *not* associative. For example:
+
+            >>> a, b = cirq.LineQubit.range(2)
+            >>> A = cirq.Circuit(cirq.H(a))
+            >>> B = cirq.Circuit(cirq.H(b))
+            >>> f = cirq.Circuit.tetris_concat
+            >>> f(f(A, B), A) == f(A, f(B, A))
+            False
+            >>> len(f(f(f(A, B), A), B)) == len(f(f(A, f(B, A)), B))
+            False
+
+        Args:
+            circuits: The circuits to concatenate.
+            stop_at_first_alignment: Defaults to false. When true, the circuits
+                are never overlapped more than needed to align their starts or
```

Comment:
```suggestion
                are never overlapped more than needed to align their starts (in case the left circuit is smaller) or
```

Reply:
This "ends (or starts) align" was hard to understand for me I tried to add a bit more explanation when these cases happen.

Reply:
@balopat it becomes jagged if you add in a different order, such as all first halves then all second halves. Neither option preserves the rectangle here:

```
from cirq import *

a, b, c = LineQubit.range(3)
c1a = Circuit(H(a), H(a), H(a))
c1b = Circuit(X(a), X(a), X(a))
c2a = Circuit(Y(b))
c2b = Circuit(Z(b), Z(b), Z(b), Z(b), Z(b))
c3a = Circuit(X(c), X(c), X(c), X(c), X(c))
c3b = Circuit(Y(c))

print(Circuit.tetris_concat(c1a, c2a, c3a, c1b, c2b, c3b, stop_at_first_alignment=False))
>> 0: ───────────H───H───H───X───X───X───
>>
>> 1: ───────────Y───Z───Z───Z───Z───Z───
>>
>> 2: ───X───X───X───X───X───Y───────────


print(Circuit.tetris_concat(c1a, c2a, c3a, c1b, c2b, c3b, stop_at_first_alignment=True))
>> 0: ───H───H───H───X───X───X───────────
>> 
>> 1: ───────────Y───Z───Z───Z───Z───Z───
>> 
>> 2: ───X───X───X───X───X───────────Y───
```

But if you change the code to initialize `upper_bound=len(c1)` the result is rectangular:
```
>> 0: ───H───H───H───X───X───X───
>> 
>> 1: ───Y───Z───Z───Z───Z───Z───
>> 
>> 2: ───X───X───X───X───X───Y───
```
This would mimic the "tetris" behavior. Though as Craig mentioned it's not symmetric. (A tetris block falling into a valley would fall all the way in, but if you did the same operation upside down, the arch (upside-down valley) would stop when its edges hit the ground, before its middle hits the block).